---
title: "The methylCC user's guide"
author: "Stephanie C. Hicks (shicks@jimmy.harvard.edu) and Rafael Irizarry (rafa@jimmy.harvard.edu)" 
date: "October 25, 2017"
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

There are several approaches available to adjust for differents in the
relative proportion of cell types in whole blood measured from
DNA methylation (DNAm). For example, *reference-based approaches*
require the use of reference data sets made up of purified cell types to
identify cell type-specific DNAm signatures. These cell type-specific DNAm
signatures are used to estimate the relative proportions of cell types
directly, but these reference data sets are laborious and expensive to
collect. Furthermore, these reference data sets will need to be
continuously collected over time as new platform technologies emerge
measuring DNAm because the observed methylation levels for the same CpGs
in the same sample vary depending the platform technology.

In contrast, there are *reference-free approaches*,
which are based on methods related to surrogate variable analysis or
linear mixed models. These approaches do not provide
estimates of the relative proportions of cell types, but rather these
methods just remove the variability induced from the differences in
relative cell type proportions in whole blood samples.

Here, we present a statistical model that estimates the cell composition
of whole blood samples measured from DNAm. The method can be applied
to microarray or sequencing data (for example whole-genome bisulfite
sequencing data, WGBS, reduced representation bisulfite sequencing
data, RRBS). Our method is based on the
idea of identifying informative genomic regions that are clearly
methylated or unmethylated for each cell type, which permits
estimation in multiple platform technologies as cell types preserve
their methylation state in regions independent of platform despite
observed measurements being platform dependent.


# Getting Started

Load the `methylCC` R package and other packages that we'll need 
later on. 

```{r loadlibs, message=FALSE, warning=FALSE}
library(methylCC)
library(minfi)
library(tidyr)
library(ggplot2)
```


# Data

## Whole Blood Illumina 450k Microarray Data Example

```{r dataload, message=FALSE}
suppressMessages(library(FlowSorted.Blood.450k))
data(FlowSorted.Blood.450k)

# Phenotypic information about samples
head(pData(FlowSorted.Blood.450k))

# RGChannelSet
rgset <- FlowSorted.Blood.450k[,
                pData(FlowSorted.Blood.450k)$CellTypeLong %in% "Whole blood"]

```


# Using the `estimateCC()` function

## Input for `estimateCC()`
The `estimateCC()` function must have
one object as input:

1. an `object` such as an `RGChannelSet` from
the R package `minfi` or a `BSseq` object
from the R package `bsseq`. This object should 
contain observed DNAm levels at CpGs (rows) 
in a set of $N$ whole blood samples (columns). 

## Running `estimateCC()`

In this example, we are interested in estimating the cell
composition of the whole blood samples listed in the
`FlowSorted.Blood.450k` R/Bioconductor package.
To run the `methylcC::estimateCC()` function,
just provide the `RGChannelSet`. This will
create an `estimateCC` object. We
will call the object `est`.


```{r run-estimateCC1}
est <- estimateCC(object = rgset, verbose = FALSE)
est
```

To see the cell composition estimates, use the
`cellcounts()` function. To see summary information 
about the object, use the `summary()` function.

```{r run-estimateCC-summaries}
cellcounts(est)
summary(est)
```


## Compare to `minfi::estimateCellCounts()` 

We can also use the `estimateCellCounts()` from R/Bioconductor package
[`minfi`](http://bioconductor.org/packages/release/bioc/html/minfi.html)
to estimate the cell composition for each of the whole blood samples. 

```{r run-minfi-estimateCellCounts}
sampleNames(rgset) <- paste0("Sample", 1:6)

countsMinfi <- minfi::estimateCellCounts(rgset)
data.frame("Tcell" = rowSums(countsMinfi[,1:3]), countsMinfi[,c(4,5,6)])
```

Then, we can compare the estimates to `methylcC::estimateCC()`. 

```{r compare-estimates}
df1 = gather(cbind("samples" = summary(est)$sampleNames,
                   data.frame("Tcell" = rowSums(countsMinfi[,1:3]), countsMinfi[,c(4,5,6)])),
             celltype, est, -samples)

df2 = gather(cbind("samples" = summary(est)$sampleNames, cellcounts(est)),
             celltype, est, -samples)

dfcombined <- dplyr::full_join(df1, df2, by = c("samples", "celltype"))

ggplot(dfcombined, aes(x=est.x, y = est.y, color = celltype)) +
    geom_point() + xlim(0,1) + ylim(0,1) +
    geom_abline(intercept = 0, slope = 1) +
    xlab("Using minfi::estimateCellCounts()") + ylab("Using methylCC::estimateCC()") +
    labs(title = "Comparing cell composition estimates")
```

We see the estimates closely match for the four cell types. 


# SessionInfo

```{r sessionInfo, results='markup'}
sessionInfo()
```

